### ISSUE TITLE
[Discussion] Pinned Rocks

### ISSUE THREAD

**Issue by vhyrro** [nvim-neorocks/rocks#102]
I'm currently implementing #86 and have encountered an interesting problem (one that I'm not sure even luarocks handles properly :p), and that's *pinning*.

When a package is pinned, all of its dependencies are subsequently also pinned. Imagine a situation like this:
```
package1
|- dependency 0.1.0

package2 (pinned)
|- dependency 0.1.0 (pinned)
```

Now, imagine you update all your rocks with `rocks update`. Since there's only one `dependency@0.1.0` directory in the rock tree, getting a broken state is incredibly easy:
```
package1
|- dependency 0.2.0

package2 (pinned)
|- dependency 0.2.0 (pinned) <- woops, it's not really pinned, is it?
```

For this reason, we should also start recognizing pinned packages and treat them differently.

There's an issue: in our rock trees, we can't differentiate using special suffixes nor prefixes, because they interfere with the version and name, respectively. My best idea would be `name@version@pinned`, but that feels incorrect.

The bottom line is: we should detect pinned dependencies and treat them as fully separate packages. In the lockfile, I hash the name, the version and also the `pinned` boolean so that pinned and unpinned packages' hashes can be differentiated easily. Making this differentiation in the rocktree is more challenging.
